### DISCUSSION TITLE
Perspective transform question

### DISCUSSION THREAD

**Discussion by ktenney** [python-pillow/Pillow#6317] [Q&A] [answered]
I am having problems with Image.transform method=Image.Transform.PERSPECTIVE
documentation seems kind of scarce.

I thought the 'data' arg was a sequence of 8 (x,y) values indicating the corners of source and destination image, however this is not working. 

Any suggestions appreciated.

Thanks,
Kent

**Comment by radarhere** [comment, marked as answer]
The official documentation is at https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.transform

When you say it's not working, what do you mean exactly? Here is code that will rotate one of our test images by 90 degrees.

```python
import math
from PIL import Image
im = Image.open("hopper.png")

angle = -math.radians(90)
matrix = [
    round(math.cos(angle), 15),
    round(math.sin(angle), 15),
    0.0,
    round(-math.sin(angle), 15),
    round(math.cos(angle), 15),
    0.0,
    0,
    0,
]
matrix[2] = (1 - matrix[0] - matrix[1]) * im.width / 2
matrix[5] = (1 - matrix[3] - matrix[4]) * im.height / 2

im.transform(im.size, Image.Transform.PERSPECTIVE, matrix).save("rotated_90.png")
```
![hopper](https://user-images.githubusercontent.com/3112309/169696773-ded2390c-dc7a-4ade-b25f-7bbdd68969f0.png)
![rotated_90](https://user-images.githubusercontent.com/3112309/169696777-0270d1cf-ebd8-4539-9389-8de95472d921.png)

**Reply by ktenney** [reply]
ok,  I was thinking of something available in cv2:
https://theailearner.com/tag/cv2-getperspectivetransform/
which takes outline points of src and dst images, returns
the correct matrix for cv2.warpPerspective

It doesn't look like Pillow offers this aid to generating a matrix.

Is the matrix returned by cv2.getPerspectiveTransform 
the same as expected by the Pillow Image.transform?

Thanks,
Kent

**Reply by radarhere** [reply]
In my earlier code, `matrix` was `[0.0, -1.0, 128.0, 1.0, 0.0, 0.0, 0, 0]`

If I ask `cv2.getPerspectiveTransform` to rotate the image,
```python
import cv2
import numpy
print(cv2.getPerspectiveTransform(
  numpy.float32([(0, 0), (128, 0), (128, 128), (0, 128)]),
  numpy.float32([(128, 0), (128, 128), (0, 128), (0, 0)])
))
```
I get
```
[[  0.  -1. 128.]
 [  1.   0.   0.]
 [ -0.   0.   1.]]
```

The only difference is that the cv2 matrix is not 8 points, but 3x3 with a 1 in the 9th slot.

**Reply by ktenney** [reply]
great, this is extremely helpful!

Thanks,
Kent
